santoku/utils/test/test_url_handler.py

```diff
@@ -1,182 +1,308 @@
-import numpy as np
-import pandas as pd
+from urllib.parse import ParseResult
+
 import pytest
-from santoku.utils.url_handler import InvalidURLError, URLHandler
+from santoku.utils.url_handler import URLHandler
+
+
+@pytest.mark.parametrize(
+    argnames=("input_subdomain", "expected_subdomain_list"),
+    argvalues=(
+        ("sub1.sub2", ["sub1", "sub2"]),
+        ("sub1", ["sub1"]),
+        ("sub1.", ["sub1", ""]),
+        ("sub1..", ["sub1", "", ""]),
+        ("", []),
+        (None, []),
+    ),
+    scope="function",
+)
+def test_string_subdomain_is_processed_as_expected(input_subdomain, expected_subdomain_list):
+    output_subdomain_list = URLHandler.process_subdomain(subdomain=input_subdomain)
+    assert output_subdomain_list == expected_subdomain_list
 
 
 @pytest.mark.parametrize(
-    argnames=("input_url", "num_subdomains", "expected_result"),
+    argnames=("input_url", "expected_components"),
     argvalues=[
-        ("https://sub2.sub1.example.com/path?query#fragment", 0, "example.com"),
-        ("https://sub2.sub1.example.com", 1, "sub1.example.com"),
-        ("https://sub2.sub1.example.com", 2, "sub2.sub1.example.com"),
-        ("https://sub2.sub1.example.com", 3, "sub2.sub1.example.com"),
-        ("https://example.com", 1, "example.com"),
-        ("https://example.co.uk", 0, "example.co.uk"),
-        ("https://user@example.co.uk:8000", 0, "example.co.uk"),
-        # Note that com.uk isn't a valid tld (but it is a valid url), so it is expected that the
+        (
+            "https://sub2.sub1.example.com/path?query#fragment",
+            ("https", ["sub2", "sub1"], "example", "com", "/path"),
+        ),
+        (
+            "https://sub1.example.com/path/",
+            ("https", ["sub1"], "example", "com", "/path/"),
+        ),
+        ("http://sub2.sub1.example.es", ("http", ["sub2", "sub1"], "example", "es", None)),
+        ("https://sub1.example.co.uk", ("https", ["sub1"], "example", "co.uk", None)),
+        ("https://example.com", ("https", [], "example", "com", None)),
+        ("https://user@example.co.uk:8000", ("https", [], "example", "co.uk", None)),
+        # Note that com.uk isn't a valid TLD (but it is a valid URL), so it is expected that the
         # funcition would consider 'example' as subdomain, 'com' as domain, and 'uk' as suffix
-        ("https://example.com.uk", 0, "com.uk"),
-        ("www.example.com", 0, "example.com"),
+        ("https://example.com.uk", ("https", ["example"], "com", "uk", None)),
```

Comment:
what's the meaning of this case?

Reply:
This test case wants to state that for `https://example.com.uk`, despite the `com.uk` is not a valid TLD, the URL is considered valid by `urllib.parse` and hence it is expected to be parsed into its components. We just want to make sure that even if a URL has an invalid TLD, it is parsed as seen here.